Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -2929,26 +2928,43 @@ private static void DetermineEqualitySupportType(Type type, ref bool thingsChang
       if (type.AsTypeParameter is { } typeArg) {
         typeArg.NecessaryForEqualitySupportOfSurroundingInductiveDatatype = true;
         thingsChanged = true;
-      } else if (type.AsMapType is { } typeMap) {
-        DetermineEqualitySupportType(typeMap.Range, ref thingsChanged);
-      } else if (type.AsSeqType is { } typeSeq) {
-        DetermineEqualitySupportType(typeSeq.Arg, ref thingsChanged);
-      } else if (type.AsIndDatatype is { } otherDt) {
-        if (otherDt.EqualitySupport == IndDatatypeDecl.ES.ConsultTypeArguments) {  // datatype is in a different SCC
-          var otherUdt = (UserDefinedType)type.NormalizeExpand();
-          var i = 0;
-          foreach (var otherTp in otherDt.TypeArgs) {
-            if (otherTp.NecessaryForEqualitySupportOfSurroundingInductiveDatatype) {
-              var tp = otherUdt.TypeArgs[i].AsTypeParameter;
-              if (tp != null) {
-                tp.NecessaryForEqualitySupportOfSurroundingInductiveDatatype = true;
-                thingsChanged = true;
+      } else if (type.SupportsEquality) {
+        // nothing to do
+      } else if (type.Normalize() is UserDefinedType userDefinedType) {
```

Comment:
Why do you need this normalization? I thought that AsMapType and AsSeqType would actually do that.

Reply:
This `type.Normalize()` cuts through any type proxies that may be sitting around the `UserDefinedType`.

`AsMapType` and `AsSeqType` call `NormalizeExpand()`, which cuts through both proxies and type synonyms. But that's too much, because a type synonym can also have an explicit `(==)` declaration.

Reply:
If a type synonym does not have an explicit `(==)` annotation, it's still sound, right?
In general, I don't think you can conclude faster that a type synonym supports equality anyway. It seems that `(==)` is only a way to catch errors earlier.

Reply:
An explicit `(==)` is also a way to advertise equality support to importers of the enclosing module. Without an explicit `(==)`, an `export provides` of the type would not let importers use the type as supporting equality.